source/iotagent-architecture.rst

```diff
@@ -0,0 +1,543 @@
+
+Certificate revocation
+----------------------
+
+An IoT agent should be able to be informed about revoked certificates. It
+should expose an API or configuration messages to allow such thing. It should
+not allow any communication with a particular device that uses a revoked
+certificate.
+
+Information context separation
+==============================
+
+A tenant could be thought simply as a group of users that share some resources.
+But its meaning goes beyond that - it implies that these resources do not share
+any common infrastructure (considering anything that transmits, processes or
+stores data) with resources belonging to other tenants. One might want to have
+separate software instances to process data from different tenants so that
+processing data from one tenant will not affect processing data from the other,
+achieving a higher level of context separation.
+
+Although this is desirable, some deployment scenarios might force using some of
+the same infrastructure for different tenants (for instance, when the
+deployment has as reduced numbers of processing units or network connections).
+So, in order to have a minimum context separation among tenants, an IoT agent
+should use everything it can to separate them, such as using different threads,
+queues, sockets, etc., and should not rely solely in deployment scenarios
+features (such as different IoT agents for different tenants). For instance,
+for topic based protocols, such as MQTT, one might want to force different
+topics for different tenants. Should a device publishes data to a particular
+topic that is owned by other tenant, this message is ignored or blocked
+(sending an error back to the device might be an optional behavior). Therefore
+no device from one tenant can send messages to any device from other tenant.
+
+The mechanism through which context separation is implemented highly depends on
+which protocol is used. A thorough analysis should be performed to properly
+implement this feature.
+
+IoT agent information and management
+====================================
+
+An IoT agent should expose all the necessary information to use it properly. It
+should expose:
+
+- **Device template**: an IoT agent should publish which is the data model it
+  accepts for a valid device. This should be done by publishing a new device
+  template to other dojot services. There should be a mechanism so that
+  different instances of the same IoT agent publishes the same device template
+  (including any template IDs). If the device template is updated in a newer
+  version of an IoT agent, the device template ID should change. The messages
+  used to publish its device template is detailed in `Device Template message`_
+  section.
+- **Management APIs**: an IoT agent should be manageable and should expose its
+  APIs to to that. The minimum set of management APIs that an IoT agent should
+  offer are:
+
+  - *Logging*: there should be a way to changing the log level of an IoT agent;
```

Comment:
to change

Reply:
isn't this feature performed by the logging library? is it really a iot agent feature? I see this feature as a generic dojot's service requirement

Reply:
Yes, it is. But it should expose an API to do that. Once we set this feature as mandatory no IoT agent should lack this feature (including those ones in Java).